doc/designs/ldap_pam_passthrough.md

```diff
@@ -0,0 +1,315 @@
+     - Otherwise continue the authentication process.
+
+ - If RADIUS is not an authentication indicator then proceed with
+   authentication.
+
+ - If OTP is an authentication indictorn or the user
+     - Evaluate tokens using the existing workflow
+     - Otherwise fall back to PASSWORD authentication
+
+OTP checking is done in ipapwd_pre_bind_otp() which is called unless
+there is a sync request. Authentication indicator type handling needs to
+be better centralized here. There are two paths that can return different
+results (assuming otpreq = False).
+
+1. User has otp auth type and has no tokens it will return 0.
+2. User does not have otp auth type it will only return 0 if the
+   user has OTP_CONFIG_AUTH_TYPE_PASSWORD.
+
+This loophole needs to be closed. Perhaps change to return false if
+auth type is OTP_CONFIG_AUTH_TYPE_NONE. This would likely be more
+future-proof if more authentication indicators are added.
+
+If the OTP check doesn't return an error then the password will be
+authenticted, if there is one. This is the two-step first check
+OTP, then check the password.
+
+Since a plugin returning 0 will allow subsequent authentication
+for the case of RADIUS we need to enforce the no password and
+radius server requirements because with PAM passthrough enabled it
+becomes a defacto default method. In fact we may want to always ensure
+there is no RADIUS server defined if the RADIUS authentication
+indicator is not set.
+
+The FATAL logging is recommended because authentication is so opaque
+administrators won't know why or where it failed. This will ensure
+a useful message is logged.
+
+It might be nice to check that PAM passthrough is enabled but there
+is is no way to validate the configuration so skip it.
+
+## Testing
+
+This creates quite a large test matrix as a number of different
+tests are required. These are only for LDAP binds. Kerberos should be
+unaffected.
+
+- RADIUS authentication indicator set and user is incorrect
+  - user has a password
+  - user has a principal key
+  - user has no radius proxy set
+
+- RADIUS authentication indicator set and user is ok
+  - test user with correct password
+  - test user with incorrect password
+
+Others depending on whether we will allow the RADIUS authentication
+indicator with others. If we restrict you either have one or not
+then this should not affect other mechanisms and will be covered by
+existing tests.
+
+### Setup
+
+In order to test we'll need to setup a RADIUS server to test against.
+The pyrad project provides a sample in
+https://raw.githubusercontent.com/pyradius/pyrad/master/example
+
+It would need to be adapted for our needs to actually do authentication
+and ideally dynamically setup its listening hosts. The passwords
+could be hardcoded with a "good" one that is always accepted.
+
+And the ldapserver PAM service needs to be created and cleaned up.
+Making a copy of system-auth is sufficient.
```

Comment:
If we use internal implementation of PAM pass-through, we don't need to manage 389-ds plugin setup and handle individually its configuration.

Reply:
done